x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java

```diff
@@ -40,18 +43,18 @@
     static final List<String> ORDERED_VALID_HOT_ACTIONS = Arrays.asList(SetPriorityAction.NAME, UnfollowAction.NAME, RolloverAction.NAME,
         ForceMergeAction.NAME);
     static final List<String> ORDERED_VALID_WARM_ACTIONS = Arrays.asList(SetPriorityAction.NAME, UnfollowAction.NAME, ReadOnlyAction.NAME,
-        AllocateAction.NAME, ShrinkAction.NAME, ForceMergeAction.NAME);
-    static final List<String> ORDERED_VALID_COLD_ACTIONS = Arrays.asList(SetPriorityAction.NAME, UnfollowAction.NAME, AllocateAction.NAME,
-        FreezeAction.NAME, SearchableSnapshotAction.NAME);
-    static final List<String> ORDERED_VALID_FROZEN_ACTIONS = Arrays.asList(SetPriorityAction.NAME, UnfollowAction.NAME, AllocateAction.NAME,
-        FreezeAction.NAME, SearchableSnapshotAction.NAME);
+        MigrateAction.NAME, AllocateAction.NAME, ShrinkAction.NAME, ForceMergeAction.NAME);
```

Comment:
I realized something about this, I think we should do `AllocateAction` *prior* to the migrate action, or else we are liable to get "stuck".

For example, if a user had multiple hot nodes and only a single warm node, we would try to migrate an index with one replica from hot to warm, but the replica could never allocate, since there is only a single node. Instead, we should do the `allocate` action first (where a user could set `number_of_replicas` to 0) before migrating to the next tier.

What do you think?

Reply:
I also realize we can get stuck the opposite way too, so either way it's possible to get stuck, but I think that doing allocate first is still the way we should go. Curious about what you think

Reply:
You make a great point Lee. I think we should go forward this way as it also keeps the "number of replicas is modified before the allocations change" behaviour consistent with how the `allocate` action works.

Maybe it's a bit more of a future talk, but I do wonder if it'd be a good time to extract the "change number of replicas" into its own action. I think being conflated with the allocation rules in the `allocate` action is generally confusing, but even more so now where we'd expect the users to use both the `allocate` and the `migrate` (admittedly, not necessarily explicitly declaring it) actions to achieve the "reduce number of replicas and then relocate index" scenario.
I think a `change-replicas-number` action would be welcome at this stage.